### ISSUE TITLE
Exterior Fuselage Shading Issue

### ISSUE THREAD

**Issue by legoboyvdlp** [c172p-team/c172p#1143]
I noticed a few shading artifacts: notice the angular lines that I circled
![image](https://user-images.githubusercontent.com/10434911/42215954-08597800-7eb8-11e8-8e4f-b8a57a2af566.png)
![image](https://user-images.githubusercontent.com/10434911/42215988-27c8c3e4-7eb8-11e8-8ce1-ad5143799cda.png)

I'm not so sure about whether one of the things I circled is an issue or not, so I put a question mark beside it

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp I believe those are mesh problems, but @wlbragg did once mention these sort of things are _very_ tricky to fix. I will let him decide if there is anything to be done about it or not.

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp This is definitely due to mesh, see:

![image](https://user-images.githubusercontent.com/5700795/42234295-0957dc7c-7eec-11e8-97cd-e08159fc3130.png)
![image](https://user-images.githubusercontent.com/5700795/42234301-0bd07130-7eec-11e8-98f3-c896ecc057ea.png)

**Comment by wlbragg** [comment]
Yes and it is so difficult to fix that and not mess something else up. I am pretty sure I actually worked on that area already and that was as good as I was going to get it. I'm taking another look at it now. I'll report back shortly.

**Comment by wlbragg** [comment]
I'm going to make more of a mess than solve if I start cutting mesh to try to remove the abnormalities. Maybe when I learn more I might be able to clean something like that up but not now. the baggage area is another area that is a mess.

**Comment by gilbertohasnofb** [comment]
@wlbragg Should we change the label of this issue for 'Future release' or do you plan to take a look at this still for this release?

**Comment by wlbragg** [comment]
Yes, change it to 'Future release'.

**Comment by gilbertohasnofb** [comment]
For future generations: if you are reading this in 10 years time, good luck with this issue! :rofl:

**Comment by legoboyvdlp** [comment]
These shading issues are characteristic of so-called n-gons. Essentially the renderer can't figure out what direction the face is pointing as n-gons have more than four vertices so it can't figure out the normal. The best method to solve it is really the knife tool to split these -- its a bit of hard work but definitely worth looking at for future releases. I'm somewhat proficient with Blender now so if you like I can take a shot at it if nobody is working on the mesh in the near future.

I would need to know what ac3d exporter to use though as that can and has caused issues especially with materials when people use different exporters

**Comment by wkitty42** [comment]
@legoboyvdlp we distribute the ac3d import/export with the craft so everyone will have and use the same one we use...it is located in the c172p's base directory with the -set file...

**Comment by legoboyvdlp** [comment]
Thanks 💯

**Comment by legoboyvdlp** [comment]
After twenty minutes work:

![image](https://user-images.githubusercontent.com/10434911/52522098-b7a12600-2c78-11e9-89b6-d54f480b6e23.png)

I've cleaned up the door mesh, so far, and started cleaning up the area around the air conditioning flap (although causing a few other issues around it)
Some un-needed vertices were actually removed on the door.

**Comment by legoboyvdlp** [comment]
And:

![image](https://user-images.githubusercontent.com/10434911/52523611-d27c9600-2c8a-11e9-88e3-cd6b33d56fd2.png)


I  think it's improved from:
![image](https://user-images.githubusercontent.com/10434911/52523633-330bd300-2c8b-11e9-9efc-96c9db3bd58e.png)

**Comment by gilbertohasnofb** [comment]
It definitely looks better in my opinion. I think this is a great moment to do this work since we are not currently doing any 3D modelling (that I am aware of), so there is no rush and we can go through the whole model. The panel in particular has some artefacts the bothered me, do you think you could take a look at those when you get a chance, lego?

Also, @wlbragg, do you think you can take a look at @legoboyvdlp 's work so far since you are much more knowledgeable in 3D modelling than me? Just to make sure that this is going in the right direction?

**Comment by gilbertohasnofb** [comment]
It definitely looks better in my opinion. I think this is a great moment to do this work since we are not currently doing any 3D modelling (that I am aware of), so there is no rush and we can go through the whole model. The panel in particular has some artefacts the bothered me, do you think you could take a look at those when you get a chance, lego?

Also, @wlbragg, do you think you can take a look at @legoboyvdlp 's work so far since you are much more knowledgeable in 3D modelling than me? Just to make sure that this is going in the right direction?

**Comment by legoboyvdlp** [comment]
Sure, if you can show me them I can take a look.

**Comment by gilbertohasnofb** [comment]
I will try to find the screenshots. They become very obvious at a very precise time of the day and heading from the sun. @wlbragg would you have any idea where I might have posted those screenshots? I don't think we ever opened an issue for that... was that maybe over the organization chat?

**Comment by wlbragg** [comment]
I just got back from my trip and will take a look. I need to refresh my memory as to all things c172p and make sure there is no open mesh work, I don't think there is.
@legoboyvdlp did you try the ac import/export we have in the repository per @wkitty42 ?

**Comment by legoboyvdlp** [comment]
Hi, 
I did -- all looks fine now, I didn't notice any broken things!

Even if there is any open mesh work I'll probably want to start again anyway -- what I did there was just mostly an experiment to show how its possible to fix the issues; I'd like to redo it and see if its possible to do it with less vertices.

**Comment by wkitty42** [comment]
remember, less vertices make for less rounded curves... there really is a trade off between having a stop sign and a circle even if the circle is really hundreds of straight lines :smile_cat:

**Comment by wlbragg** [comment]
>Also, @wlbragg, do you think you can take a look at @legoboyvdlp 's work so far since you are much more knowledgeable in 3D modelling than me? Just to make sure that this is going in the right direction?

@gilbertohasnofb once @legoboyvdlp gets to a final PR version, I'll take a look. I have no doubt it can be cleaned up, it was simply more than I wanted to deal with, especially the liveries behind the mesh. After what I did to the nose cowling, wing tips, tail section and all the really big issues I noticed,  I pretty much had my fill of mesh cleanup. I'm glad someone else can follow me up and get to some of the details.
The entire fuselage could still stand a remake but we have come a long way from where we were!

**Comment by legoboyvdlp** [comment]
I don't think the last person who exported c172p.ac used our exporter

our exporter export_ac3d.py:

> self.world = AC3D.World('Blender_export__' + bpy.path.basename(filepath), self.export_conf)

this one:
https://github.com/NikolaiVChr/Blender-AC3D
>self.world = AC3D.World('Blender_exporter_v' +str(operator.v_info[0])+"."+str(operator.v_info[1])+"__" + bpy.path.basename(filepath), self.export_conf)


our .ac has `name "Blender_exporter_v2.26__c172p.ac"` while when I exported using our exporter I got `Blender_export__c172p.ac`.

since it looks like we actually use the one on NikolaiVChr's github based on the name of the exported .ac, I'll use that one.

**Comment by gilbertohasnofb** [comment]
> The entire fuselage could still stand a remake but we have come a long way from where we were!

Absolutely! This model is light-years ahead of where it was just a couple of years ago!

![image](https://user-images.githubusercontent.com/5700795/52714399-aa927880-2f91-11e9-8ccf-a09ee852954e.png)

![image](https://user-images.githubusercontent.com/5700795/52714406-acf4d280-2f91-11e9-9ced-611c4ad872af.png)

**Comment by wkitty42** [comment]
@legoboyvdlp might need to check with @wlbragg ? maybe he has an updated version? he's been the main one doing the .ac files AFAIK... at least the most prolific... if there is an updated version, it should be updated in the repo and fgdata...

**Comment by legoboyvdlp** [comment]
I used the exporter I linked and everything worked fine: no broken materials. When I used the one in my repo, all the materials broke.

**Comment by wlbragg** [comment]
@legoboyvdlp  I am guilty of using different .ac's I'm sure at one time I used the one in the repo, since then I use whichever one I find that works across most of the different aircraft I am working on at the time. The Shuttle is really picky and needs a particular version. I have discovered the usual issues are Grouping, either it works or doesn't, and the way the materials are exported can be an issue, (as with the Shuttle) and the exporter you originally used.
I too would just as soon keep up with NikolaiVChr's github based version if possible. I need to give it a go on the shuttle again and see if I can troubleshoot the issue it has with the Shuttle (materials) and have NikolaiVChr patch the latest version, or at least understand what the issue is. Whatever works correctly is fine with me. Anyone working on the mesh needs to be proficient enough to understand the import/export nuances which I'm sure your are.

**Comment by wlbragg** [comment]
Yes, @wkitty42 is technically correct, we should update the repo with the most current io script available that works.

**Comment by legoboyvdlp** [comment]
I think we'll keep this open, I'm not yet happy with the shading. But I don't think I'll have the time to do any more work unfortunately until this summer - probably July and August.
